### PR TITLE
Copyable concern

### PR DESCRIPTION
**Pull request by mbandrewfoster**

### CODE REVIEW

File: app/models/concerns/nunchaku/copyable.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      end
+
+      def copyable_associations
+        []
+      end
+    end
+
+    def copy(deep=false)
+      build_copy(self.class, self, self).tap { |c| copy_associations(c) if deep }
+    end
```
Comment (ozvillafan) [inline app/models/concerns/nunchaku/copyable.rb]: def copy
 build_copy(self.class, self, self)
end

def deep_copy
  copy.tap { |c| copy_associations(c) }
end

Comment (mbandrewfoster) [inline app/models/concerns/nunchaku/copyable.rb]: Yep, that's better. Updated.

File: app/models/concerns/nunchaku/copyable.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    def copied_attributes(klass)
+      attributes.reject {|k, v| klass.copyable_attributes_blacklist.include?(k.to_sym) }
+    end
+
+    protected
+
+    def build_copy(klass, source, parent)
+      klass.new(source.copied_attributes(klass)).tap { |r| r.assign_attributes(klass.copyable_overridden_attributes(source, parent)) }
```
Comment (ozvillafan) [inline app/models/concerns/nunchaku/copyable.rb]: I think this works

klass.new(source.copied_attributes(klass).merge(klass.copyable_overridden_attributes(source,parent))

File: app/models/concerns/nunchaku/copyable.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      end
+    end
+
+    def copied_attributes(klass)
+      attributes.reject {|k, v| klass.copyable_attributes_blacklist.include?(k.to_sym) }
+    end
+
+    protected
+
+    def build_copy(klass, source, parent)
```
Comment (ozvillafan) [inline app/models/concerns/nunchaku/copyable.rb]: I'm not sure what parent does?

Comment (mbandrewfoster) [inline app/models/concerns/nunchaku/copyable.rb]: Nothing right now, but I had a thought that the copyable_overridden_attributes may need to know something about the parent. I'll take it out, will leave it simpler for now, can always add it back later.

**Pull request merged**